File: lib/wallet_key_validator_kit.dart
import 'package:crypto/crypto.dart';

const _alphabet = '123456789ABCDEFGHJKLMNPQRSTUVWXYZabcdefghijkmnopqrstuvwxyz';
const _secpOrder = 'fffffffffffffffffffffffffffffffebaaedce6af48a03bbfd25e8cd0364141';
const _wif = {'bitcoin': 0x80, 'testnet': 0xef, 'litecoin': 0xb0, 'dogecoin': 0x9e, 'dash': 0xcc, 'bitcoin-cash': 0x80, 'zcash': 0x80, 'digibyte': 0x80};
const _ext = {'0488b21e': 'xpub', '0488ade4': 'xprv', '049d7cb2': 'ypub', '049d7878': 'yprv', '04b24746': 'zpub', '04b2430c': 'zprv', '043587cf': 'tpub', '04358394': 'tprv', '044a5262': 'upub', '044a4e28': 'uprv', '045f1cf6': 'vpub', '045f18bc': 'vprv', '019da462': 'Ltub', '019d9cfe': 'Ltpv', '01b26ef6': 'Mtub', '01b26792': 'Mtpv'};

Map<String, Object?> validatePrivateKey(String value, {String chain = 'bitcoin'}) {
  final body = (value.startsWith('0x') || value.startsWith('0X')) ? value.substring(2) : value;
  if (chain == 'solana') return {'valid': RegExp(r'^([0-9a-fA-F]{64}|[0-9a-fA-F]{128})$').hasMatch(body), 'format': 'ed25519-hex'};
  final valid = RegExp(r'^[0-9a-fA-F]{64}$').hasMatch(body) && body.contains(RegExp(r'[1-9a-fA-F]')) && body.toLowerCase().compareTo(_secpOrder) < 0;
  return {'valid': valid, 'format': 'secp256k1-hex'};
}

Map<String, Object?> validateWif(String value, {String chain = 'bitcoin'}) {
  final payload = _base58Check(value.trim());
  if (payload == null) return {'valid': false, 'format': 'wif', 'reason': 'bad checksum'};
  final compressed = payload.length == 34 && payload.last == 1;
  return {'valid': payload.first == _wif[chain] && (payload.length == 33 || compressed), 'format': 'wif', 'compressed': compressed};
}

Map<String, Object?> validateExtendedKey(String value) {
  final payload = _base58Check(value.trim());
  if (payload == null || payload.length != 78) return {'valid': false, 'format': 'extended-key'};
  final version = payload.take(4).map((b) => b.toRadixString(16).padLeft(2, '0')).join();
  return {'valid': _ext.containsKey(version), 'format': 'extended-key', 'prefix': _ext[version]};
}

List<int>? _base58Decode(String value) {
  var bytes = <int>[0];
  for (final unit in value.runes) {
    final digit = _alphabet.indexOf(String.fromCharCode(unit));
    if (digit < 0) return null;
    var carry = digit;
    for (var i = bytes.length - 1; i >= 0; i--) { final n = bytes[i] * 58 + carry; bytes[i] = n & 0xff; carry = n >> 8; }
    while (carry > 0) { bytes.insert(0, carry & 0xff); carry >>= 8; }
  }
  final zeros = value.runes.takeWhile((r) => r == '1'.codeUnitAt(0)).length;
  return [...List<int>.filled(zeros, 0), ...bytes.skipWhile((b) => b == 0)];
}

List<int>? _base58Check(String value) {
  final raw = _base58Decode(value);
  if (raw == null || raw.length < 5) return null;
  final payload = raw.sublist(0, raw.length - 4);
  final checksum = raw.sublist(raw.length - 4);
  final second = sha256.convert(sha256.convert(payload).bytes).bytes;
  for (var i = 0; i < 4; i++) { if (second[i] != checksum[i]) return null; }
  return payload;
}

File: Sources/WalletKeyValidatorKit/WalletKeyValidatorKit.swift
import Foundation
import CryptoKit

public struct KeyValidationResult: Equatable {
    public let valid: Bool
    public let format: String?
    public let prefix: String?
    public let reason: String?
}

public enum WalletKeyValidatorKit {
    private static let alphabet = Array("123456789ABCDEFGHJKLMNPQRSTUVWXYZabcdefghijkmnopqrstuvwxyz")
    private static let secpOrder = "fffffffffffffffffffffffffffffffebaaedce6af48a03bbfd25e8cd0364141"
    private static let wif: [String: UInt8] = ["bitcoin": 0x80, "testnet": 0xef, "litecoin": 0xb0, "dogecoin": 0x9e, "dash": 0xcc, "bitcoin-cash": 0x80, "zcash": 0x80, "digibyte": 0x80]
    private static let ext: [String: String] = ["0488b21e": "xpub", "0488ade4": "xprv", "049d7cb2": "ypub", "049d7878": "yprv", "04b24746": "zpub", "04b2430c": "zprv", "043587cf": "tpub", "04358394": "tprv", "044a5262": "upub", "044a4e28": "uprv", "045f1cf6": "vpub", "045f18bc": "vprv", "019da462": "Ltub", "019d9cfe": "Ltpv", "01b26ef6": "Mtub", "01b26792": "Mtpv"]

    public static func validatePrivateKey(_ value: String, chain: String = "bitcoin") -> KeyValidationResult {
        let body = value.hasPrefix("0x") || value.hasPrefix("0X") ? String(value.dropFirst(2)) : value
        if chain == "solana" {
            let valid = (body.count == 64 || body.count == 128) && body.allSatisfy(\.isHexDigit)
            return KeyValidationResult(valid: valid, format: "ed25519-hex", prefix: nil, reason: nil)
        }
        let valid = body.count == 64 && body.allSatisfy(\.isHexDigit) && body.contains(where: { $0 != "0" }) && body.lowercased() < secpOrder
        return KeyValidationResult(valid: valid, format: "secp256k1-hex", prefix: nil, reason: nil)
    }

    public static func validateWif(_ value: String, chain: String = "bitcoin") -> KeyValidationResult {
        guard let payload = base58Check(value.trimmingCharacters(in: .whitespacesAndNewlines)) else {
            return KeyValidationResult(valid: false, format: "wif", prefix: nil, reason: "bad checksum")
        }
        let compressed = payload.count == 34 && payload.last == 1
        let valid = wif[chain].map { payload.first == $0 && (payload.count == 33 || compressed) } ?? false
        return KeyValidationResult(valid: valid, format: "wif", prefix: nil, reason: nil)
    }

    public static func validateExtendedKey(_ value: String) -> KeyValidationResult {
        guard let payload = base58Check(value.trimmingCharacters(in: .whitespacesAndNewlines)), payload.count == 78 else {
            return KeyValidationResult(valid: false, format: "extended-key", prefix: nil, reason: "invalid extended key")
        }
        let version = payload.prefix(4).map { String(format: "%02x", $0) }.joined()
        let prefix = ext[version]
        return KeyValidationResult(valid: prefix != nil, format: "extended-key", prefix: prefix, reason: nil)
    }

    private static func base58Decode(_ value: String) -> [UInt8]? {
        var bytes: [UInt8] = [0]
        for character in value {
            guard let digit = alphabet.firstIndex(of: character) else { return nil }
            var carry = digit
            for index in stride(from: bytes.count - 1, through: 0, by: -1) {
                let n = Int(bytes[index]) * 58 + carry
                bytes[index] = UInt8(n & 0xff)
                carry = n >> 8
            }
            while carry > 0 { bytes.insert(UInt8(carry & 0xff), at: 0); carry >>= 8 }
        }
        return Array(repeating: 0, count: value.prefix(while: { $0 == "1" }).count) + Array(bytes.drop(while: { $0 == 0 }))
    }

    private static func base58Check(_ value: String) -> [UInt8]? {
        guard let raw = base58Decode(value), raw.count >= 5 else { return nil }
        let payload = Array(raw.dropLast(4))
        let checksum = Array(raw.suffix(4))
        let first = SHA256.hash(data: Data(payload))
        let second = SHA256.hash(data: Data(first))
        return Array(second.prefix(4)) == checksum ? payload : nil
    }
}
